doc/adrs/057_add_support_for_tool_precondition.adoc

```diff
@@ -0,0 +1,31 @@
+= ADR-057 - Add support for tool preconditions
+
+== Context
+
+As of now, to check if a tool can be applied on a diagram element (as a source or target of the tool), we check that its diagram description is within the candidates of the tool.
+The semantic object represented by the diagram element does not play a role in this checking process.
+Only when the tool is applied do we know if the tool will successfully apply to the semantic object.
+
+== Decision
+
+We will add an way for a tool to accept a semantic object as a source or target in the diagram tools interfaces `SingleClickOnTwoDiagramElementsTool` and `SingleClickOnDiagramElementTool`.
+The connector tool will check against the backend if the tools it is given are valid for the chosen source and target.
+
+=== Backend
+
+The `SingleClickOnTwoDiagramElementsTool` and `SingleClickOnDiagramElementTool` will have new fields containing functions that can validate a semantic source/target.
```

Comment:
If I don't add those fields on SingleClickOnTwoDiagramElementsTool what is the proper location to allow the implementation of preconditions ?

I haven't seen a way to put a precondition on the semantic element for a SingleClickOnDiagramElementTool but it is indeed irrelevant to the connector tools, I added it just in case.

Reply:
Look at `CompatibilityToolSectionsProvider` to see how we are leveraging today the Sirius Desktop preconditions in Sirius Components without having to introduce a precondition concept in Sirius Components itself. Sirius Components asks for the tools that are relevant on a given object. It does not care how this computation is performed.